src/vma/dev/ring_eth_cb.cpp

```diff
@@ -0,0 +1,314 @@
+
+	// create cyclic buffer get exception on failure
+	alloc.alloc_and_reg_mr(m_buffer_size, p_ring_info->p_ib_ctx) ;
+	ring_simple::create_resources(p_ring_info, active);
+	m_is_mp_ring = true;
+	m_ibv_rx_sg_array = m_p_qp_mgr->get_rx_sge();
+	ring_logdbg("use buffer parameters, m_buffer_size %zd "
+		    "strides_num %d stride size %d",
+		    m_buffer_size, m_strides_num, m_stride_size);
+}
+
+qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
+				   uint8_t port_num,
+				   struct ibv_comp_channel *p_rx_comp_event_channel) throw (vma_error)
+{
+	return new qp_mgr_mp(this, ib_ctx, port_num, p_rx_comp_event_channel,
+			get_tx_num_wr(), get_partition());
+}
+
+int ring_eth_cb::drain_and_proccess(cq_type_t cq_type)
+{
+	NOT_IN_USE(cq_type);
+	return 0;
+}
+
+int ring_eth_cb::poll_and_process_element_rx(uint64_t* p_cq_poll_sn,
+					     void* pv_fd_ready_array)
+{
+	NOT_IN_USE(p_cq_poll_sn);
+	NOT_IN_USE(pv_fd_ready_array);
+	return 0;
+}
+
+int ring_eth_cb::cyclic_buffer_read(vma_completion_mp_t &completion,
+				    size_t min, size_t max, int &flags)
+{
+	uint32_t offset = 0, p_flags = 0;
+	uint16_t size = 0;
+	volatile struct mlx5_cqe64 *cqe64;
+
+	// sanity check
+	if (unlikely(min > max || max == 0 || flags != MSG_DONTWAIT)) {
+		errno = EINVAL;
+		ring_logdbg("Illegal values, got min: %d, max: %d, flags %d",
+			    min, max, flags);
+		if (flags != MSG_DONTWAIT) {
+			ring_logdbg("only %d flag is currently supported",
+				    MSG_DONTWAIT);
+		}
+		return -1;
+	}
+
+	int ret = ((cq_mgr_mp *)m_p_cq_mgr_rx)->poll_mp_cq(size,
+					m_stride_counter, offset, p_flags, cqe64);
+	// empty
+	if (size == 0) {
+		return 0;
+	}
+	if (unlikely(ret == -1)) {
+		ring_logdbg("poll_mp_cq failed with errno %m", errno);
+		return -1;
+	}
+	// set it here because we might not have min packets avail in this run
+	if (likely(!(p_flags & VMA_MP_RQ_BAD_PACKET))) {
+		if (unlikely(m_curr_d_addr == 0)) {
+			m_curr_d_addr = (void *)(m_ibv_rx_sg_array[m_curr_wq].addr + offset);
+			if (completion.comp_mask & VMA_MP_MASK_TIMESTAMP) {
+				convert_hw_time_to_system_time(ntohll(cqe64->timestamp),
+							       &m_curr_hw_timestamp);
+			}
+			// When UMR will be added this will be different
+			m_curr_h_ptr = m_curr_d_addr;
+			m_curr_packets = 1;
+			m_curr_size = size;
+		} else {
+			m_curr_packets++;
+			m_curr_size += size;
+		}
+		if (unlikely(m_stride_counter >= m_pow_strides_num)) {
+			reload_wq();
+		} else {
+			ret = mp_loop(min);
+			if (ret == 1) { // there might be more to drain
+				mp_loop(max);
+			} else if (ret == 0) { // no packets left
+				return 0;
+			}
+		}
+	}
+
+	completion.payload_ptr = m_curr_d_addr;
+	completion.payload_length = m_curr_size;
+	completion.packets = m_curr_packets;
+	if (completion.comp_mask & VMA_MP_MASK_HDR_PTR) {
+		completion.headers_ptr = m_curr_h_ptr;
+		completion.headers_ptr_length = m_curr_size;
+	}
+	completion.hw_timestamp = m_curr_hw_timestamp;
```

Comment:
Add a comment explaining that `m_curr_hw_timestamp` is set for that of the first packet only.